cache/expirationcache/prefetching_cache.go

```diff
@@ -0,0 +1,127 @@
+package expirationcache
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+type PrefetchingExpiringLRUCache[T any] struct {
+	cache                   ExpiringCache[cacheValue[T]]
+	prefetchingNameCache    ExpiringCache[atomic.Uint32]
+	reloadFn                ReloadEntryFn[T]
+	prefetchThreshold       int
+	prefetchExpires         time.Duration
+	onPrefetchEntryReloaded OnEntryReloadedCallback
+	onPrefetchCacheHit      OnCacheHitCallback
+}
+
+type cacheValue[T any] struct {
+	element  *T
+	prefetch bool
+}
+
+// OnEntryReloadedCallback will be called if a prefetched entry is reloaded
+type OnEntryReloadedCallback func(key string)
+
+// ReloadEntryFn reloads a prefetched entry by key
+type ReloadEntryFn[T any] func(key string) (*T, time.Duration)
+
+type PrefetchingOptions[T any] struct {
+	Options
+	ReloadFn                func(cacheKey string) (*T, time.Duration)
+	PrefetchThreshold       int
+	PrefetchExpires         time.Duration
+	PrefetchMaxItemsCount   int
+	OnPrefetchAfterPut      OnAfterPutCallback
+	OnPrefetchEntryReloaded OnEntryReloadedCallback
+	OnPrefetchCacheHit      OnCacheHitCallback
+}
+
+type PrefetchingCacheOption[T any] func(c *PrefetchingExpiringLRUCache[cacheValue[T]])
+
+func NewPrefetchingCache[T any](options PrefetchingOptions[T]) *PrefetchingExpiringLRUCache[T] {
+	pc := &PrefetchingExpiringLRUCache[T]{
+		prefetchingNameCache: NewCache[atomic.Uint32](Options{
+			CleanupInterval: time.Minute,
+			MaxSize:         uint(options.PrefetchMaxItemsCount),
+			OnAfterPutFn:    options.OnPrefetchAfterPut,
+		}),
+		prefetchExpires:         options.PrefetchExpires,
+		prefetchThreshold:       options.PrefetchThreshold,
+		reloadFn:                options.ReloadFn,
+		onPrefetchEntryReloaded: options.OnPrefetchEntryReloaded,
+		onPrefetchCacheHit:      options.OnPrefetchCacheHit,
+	}
+
+	pc.cache = NewCacheWithOnExpired[cacheValue[T]](options.Options, pc.onExpired)
+
+	return pc
+}
+
+// check if a cache entry should be prefetched: was queried > threshold in the time window
+func (e *PrefetchingExpiringLRUCache[T]) shouldPrefetch(cacheKey string) bool {
+	if e.prefetchThreshold == 0 {
+		return true
+	}
+
+	cnt, _ := e.prefetchingNameCache.Get(cacheKey)
+
+	return cnt != nil && int(cnt.Load()) > e.prefetchThreshold
```

Comment:
Even though it's unlikely an entry's counter would be `> IntMax` might as well use `int64` to guarantee we can represent all `uint32` values:
```suggestion
	return cnt != nil && int64(cnt.Load()) > e.prefetchThreshold
```